crates/circuit/src/dag_circuit.rs

```diff
@@ -6501,6 +6509,184 @@ impl DAGCircuit {
+            new_dag.calibrations = calibrations;
+        }
+
+        new_dag.metadata = qc.metadata.map(|meta| meta.unbind());
+
+        // Add the qubits depending on order.
+        let mut qubit_map: HashMap<Qubit, Bound<PyAny>> = HashMap::new();
+        if let Some(qubit_ordering) = qubit_order {
+            for qubit in qubit_ordering {
+                if new_dag.qubits.find(&qubit).is_some() {
+                    return Err(DAGCircuitError::new_err(format!(
+                        "duplicate qubits {}",
+                        &qubit
+                    )));
+                }
+                qubit_map.insert(new_dag.add_qubit_unchecked(py, &qubit)?, qubit);
+            }
+        } else {
+            for qubit in qc_data.qubits().bits() {
+                let bound = qubit.clone_ref(py).into_bound(py);
+                qubit_map.insert(new_dag.add_qubit_unchecked(py, qubit.bind(py))?, bound);
+            }
+        }
+
+        // Add the clbits depending on order.
+        let mut clbit_map: HashMap<Clbit, Bound<PyAny>> = HashMap::new();
+        if let Some(clbit_ordering) = clbit_order {
+            for clbit in clbit_ordering {
+                if new_dag.clbits.find(&clbit).is_some() {
+                    return Err(DAGCircuitError::new_err(format!(
+                        "duplicate clbits {}",
+                        &clbit
+                    )));
+                }
+                clbit_map.insert(new_dag.add_clbit_unchecked(py, &clbit)?, clbit);
+            }
+        } else {
+            for clbit in qc_data.clbits().bits() {
+                let bound = clbit.clone_ref(py).into_bound(py);
+                clbit_map.insert(new_dag.add_clbit_unchecked(py, clbit.bind(py))?, bound);
+            }
+        }
+
+        // Add all of the new vars.
+        for var in &qc.declared_vars {
+            new_dag.add_var(py, var, DAGVarType::Declare)?;
+        }
+
+        for var in &qc.input_vars {
+            new_dag.add_var(py, var, DAGVarType::Input)?;
+        }
+
+        for var in &qc.captured_vars {
+            new_dag.add_var(py, var, DAGVarType::Capture)?;
+        }
+
+        // Add all the registers
+        if let Some(qregs) = qc.qregs {
+            for qreg in qregs.iter() {
+                new_dag.add_qreg(py, &qreg)?;
+            }
+        }
+
+        if let Some(cregs) = qc.cregs {
+            for creg in cregs.iter() {
+                new_dag.add_creg(py, &creg)?;
+            }
+        }
+
+        // Collect the re-mapped indices for qubits
+        let remapped_qubits: HashMap<Qubit, Qubit> = qubit_map
+            .into_iter()
+            .map(|(index, bit)| (qc_data.qubits().find(&bit).unwrap(), index))
+            .collect();
+        // Collect the re-mapped indices for clbits
+        let remapped_clbits: HashMap<Clbit, Clbit> = clbit_map
+            .into_iter()
+            .map(|(index, bit)| (qc_data.clbits().find(&bit).unwrap(), index))
+            .collect();
```

Comment:
I'm a bit concerned about this and the hashmap above in the default no order arguments being specified. We're spending a fair amount of effort building mappings of basically `{0: 0, 1:1, ..., n: n}` and we should just short circuit that path without any allocations as this becomes unnecessary overhead in those cases. Also is there a reason to not use a `Vec` here for these two? The `Qubit` and `Clbit` are indices into a vec, so they always need to be `0..n`?

Reply:
I removed this second re-mapping, and made everything happen in the first one, so that the outcome is optional if no ordering is provided.